### PR TITLE
Multi-channel Support

### PR DESCRIPTION
**Pull request by Carlson-J**
These changes allow for additional channels to be loaded/created when an image is loaded, e.g., from the metadata in the image. The behavior should be unchanged for 1-3 channel data.

### Summary
🌈🎨🛠️

This pull request enhances the `ultralytics/yolo` package to support images with more than three channels, such as alpha or depth. It updates the image augmentation, validation, and export modules to handle different numbers of input channels, and modifies the plotting module to only plot the RGB channels.

> _Sing, O Muse, of the cunning code reviewers_
> _Who scrutinized the changes of the Exporter_
> _And made it heed the channels of the models_
> _As varied as the hues of Iris' bow._

### Walkthrough
*  Support models with different numbers of input channels than the default 3 ([link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-89bbf0a641e1fc4302b73181ac7fccc773b5213720cae9b5cae8aa2393c982fdL174-R175), [link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-7646fc7a670fe6bbcbf9e637a707139b362bd3b118be7faca28bff7d9b7450c3R22), [link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-7646fc7a670fe6bbcbf9e637a707139b362bd3b118be7faca28bff7d9b7450c3R91-R109), [link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-7646fc7a670fe6bbcbf9e637a707139b362bd3b118be7faca28bff7d9b7450c3L139-R163))
* Modify augmentation classes in `augment.py` to only apply transformations to the first three channels of the image, which are assumed to be BGR, and preserve any additional channels ([link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-13d355434447b3961ec61d1c80bedc2b6db2d98ab0df572eb036f4e8b536aee9L403-R403), [link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-13d355434447b3961ec61d1c80bedc2b6db2d98ab0df572eb036f4e8b536aee9L412-R412), [link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-13d355434447b3961ec61d1c80bedc2b6db2d98ab0df572eb036f4e8b536aee9L489-R489), [link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-13d355434447b3961ec61d1c80bedc2b6db2d98ab0df572eb036f4e8b536aee9L585-R586), [link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-13d355434447b3961ec61d1c80bedc2b6db2d98ab0df572eb036f4e8b536aee9L645-R646))
* Modify `plot_images` function in `plotting.py` to only use the first three channels of the images, which are assumed to be RGB, for plotting ([link](https://github.com/ultralytics/ultralytics/pull/1701/files?diff=unified&w=0#diff-7e75b3f968b9868ef6f0a86148cf6c16a5224e7fdd45f5c2eb5e9bc0fc744756R247))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced image augmentations and model input flexibility in Ultralytics' YOLO 🖼️✨.

### 📊 Key Changes
- **Adapted HSV conversion:** Ensures compatibility with images not strictly in the BGR format by limiting operation to the first three channels.
- **Border addition method:** Improved by using NumPy's `np.pad` for efficiency and simplicity.
- **Channel handling in transformations:** Adjusted to properly handle images with varying numbers of channels beyond the basic RGB.
- **Dynamic input channel support:** Modifications to allow models to adapt to input images with different channel numbers, enhancing model flexibility.
- **Error handling for model warming up:** Added logic to dynamically adjust to the required number of input channels based on potential runtime errors.
- **Image plotting adjustments:** Ensured only RGB channels are considered during plotting to avoid errors with images having extra channels.

### 🎯 Purpose & Impact
- **For developers and researchers:** These changes make the YOLO model more versatile and easier to work with, especially when dealing with images that have more than just the standard RGB or BGR channels.
- **For users:** Improves the accuracy and robustness of image detection, particularly in scenarios where images may come from varied sources with different formatting.
- **Overall:** The modifications lead to a more flexible architecture, capable of handling a broader range of image types and enhancing the model's usability across diverse datasets.

### CODE REVIEW

**Comment by Carlson-J** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Carlson-J** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@Carlson-J thanks for the PR! We automatically extend greyscale images to 3-ch, as well as convert 4-ch PNGs with alpha channel to 3-ch.

HSV should have an `if` to skip if not compatible with the incoming images, but I think wider changes should probably be part of a larger conversation around the changes required for higher channel (i.e. hyperspectral imagery)

**Comment by Carlson-J** [comment]
@glenn-jocher My specific case for wanting multi-channel input is to propagate additional metadata data about the images as feature channels. I plan on adding it in at future stages, e.g., at each of the detection heads, but it is not clear how to do this when using mosaic. It is also the case that these features can vary spatially, so accounting for that would be easiest at the first layer. 

I understand that in the case of multi-spectral data, using the HSV is not wanted, but in the case where features are being appended to images, it makes sense to only do HSV augmentation on the first three channels. I'm willing to implement these features if that is a direction you are willing to take this framework, i.e., incorporating a richer set of features/metadata into the images.

**Comment by glenn-jocher** [comment]
Thanks for sharing those details, @Carlson-J. It sounds like using additional channels to represent metadata data about images as features would be a useful addition. While we want to avoid breaking changes in the framework and ensure compatibility with various image types and use cases, we are open to discussing potential changes and considering them for implementation as part of a larger conversation.

**Comment by Carlson-J** [comment]
@glenn-jocher Ok, how do we go about having this larger conversation?

**Comment by glenn-jocher** [comment]
@Carlson-J Great question! We can start by creating an issue in the repository to discuss this feature and its potential use cases. The issue can be used to gather ideas, discuss possible solutions, and get feedback from other members of the community. Additionally, we can use the Issue to outline a roadmap that aligns with the long-term goals of the framework.

**Comment by yjmd2222** [comment]
Please excuse my ignorance, but what is the status of this PR? Can I use this on 4-channel images? I'm confused because @glenn-jocher  you said 

> We automatically extend greyscale images to 3-ch, as well as convert 4-ch PNGs with alpha channel to 3-ch.

Does this mean all images are converted to 3-ch images at this stage?

I see no apparent difference in the model summary when I feed in a 4-ch png file.
```
                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.Conv                  [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.Conv                  [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.C2f                   [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.Conv                  [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.C2f                   [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.Conv                  [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.C2f                   [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.Conv                  [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.C2f                   [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.SPPF                  [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.Concat                [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.C2f                   [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.Concat                [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.C2f                   [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.Conv                  [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.Concat                [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.C2f                   [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.Conv                  [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.Concat                [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.C2f                   [384, 256, 1]                 
 22        [15, 18, 21]  1    751702  ultralytics.nn.modules.Detect                [2, [64, 128, 256]]           
Model summary: 225 layers, 3011238 parameters, 3011222 gradients, 8.2 GFLOPs
```
Thanks.

**Comment by Phoenix2022go** [comment]
Thank you for your contribution @Carlson-J.
For now, I am facing the multi-channels image detect problem: I have a training set consist of pixel aligned RGB and IR images and the image format is "jpg". 
Could you please tell me if your contribution been packaged in the official files? Could you please tell me how to use your method as mentioned earlier to achieve the multi-channel image detection task?
